types/logger.d.ts

```diff
@@ -19,15 +19,19 @@ export type Bindings = pino.Bindings
 
 export type ChildLoggerOptions = pino.ChildLoggerOptions
 
-export type FastifyBaseLogger = pino.BaseLogger & {
-  child(bindings: Bindings, options?: ChildLoggerOptions): FastifyBaseLogger
+export interface FastifyBaseLogger<T extends FastifyBaseLogger<any> = FastifyBaseLogger<any>> extends pino.BaseLogger {
+  child(bindings: Bindings, options?: ChildLoggerOptions): T
+}
+
+export interface DefaultFastifyLogger extends FastifyBaseLogger<DefaultFastifyLogger>, Pick<pino.Logger, 'setBindings'> {
+
```

Comment:
FastifyBaseLogger.child needs to return the DefaultFastifyLogger type, and there is a circular dependency problem if it is written like this.
```ts
export type FastifyBaseLogge = pino.BaseLogger & {
  child(bindings: Bindings, options?: ChildLoggerOptions): DefaultFastifyLogger
}

export interface DefaultFastifyLogger extends FastifyBaseLogger<DefaultFastifyLogger>, Pick<pino.Logger, 'setBindings'> {
```

So I used CRTP (curiously recurring template pattern) here

Reply:
return `this` should be possible.

Reply:
I tried, but there were problems in some cases. See:

<img width="1339" alt="image" src="https://user-images.githubusercontent.com/8198408/196121604-34ef9fc7-9dcc-41eb-9f7d-1618b32d5574.png">